### PR TITLE
MODEXPW-233 - Migrate MinIO adapter changes to master branch

### PR DESCRIPTION
**Pull request by khandramai**
[MODEXPW-233](https://issues.folio.org/browse/MODEXPW-233) - Migrate MinIO adapter changes to master branch

## Purpose
MG code migration to master branch

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by khandramai** [comment]
@SerhiiNosko , could you please assist in review of this PR?

**Comment by SerhiiNosko** [comment]
> @SerhiiNosko , could you please assist in review of this PR?

Sure, will do

**Review by alekGbuz — APPROVED** [review]

File: src/main/java/org/folio/dew/controller/BulkEditController.java [hunk 1]
```diff
@@ -482,10 +476,10 @@ private String buildPreviewQueryFromCsv(JobCommand jobCommand, int limit) {
     try {
       Reader inputReader;
       var minioFileName = PREVIEW_PREFIX + FilenameUtils.getName(fileName);
-      if (Files.notExists(Path.of(fileName)) && repository.containsFile(minioFileName)) {
-        inputReader = new InputStreamReader(repository.getObject(minioFileName));
+      if (localFilesStorage.notExists(fileName) && remoteFilesStorage.containsFile(minioFileName)) {
+        inputReader = new InputStreamReader(remoteFilesStorage.newInputStream(minioFileName));
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dew/controller/BulkEditController.java:480]: You are using try with resources in this place: https://github.com/folio-org/mod-data-export-worker/pull/309/files#diff-39c0f2fcfdeae51e6f2d49edeb58949c937f4026f59e7c6f34ce55b453dabfd9R278
Do you need to use the same here?

File: src/main/java/org/folio/dew/repository/BaseFilesStorage.java [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+      var sb = new StringBuilder();
+
+      paths.forEach(p -> {
+        try {
+          client.removeObject(RemoveObjectArgs.builder()
+            .bucket(bucket)
+            .region(region)
+            .object(p)
+            .build());
+        } catch (Exception e) {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dew/repository/BaseFilesStorage.java:280]: I think better to add error message in log to understand why we could not delete file

File: src/main/java/org/folio/dew/repository/BaseFilesStorage.java [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+        .bucket(bucket)
+        .region(region)
+        .prefix(path)
+        .maxKeys(1)
+        .build())
+      .iterator();
+    try {
+      return iterator.hasNext() && Objects.nonNull(iterator.next().get());
+    } catch (Exception e) {
+      log.error("Error file existing verification, path: " + path);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dew/repository/BaseFilesStorage.java:321]: The same here, we can add error message from exception into existing log

File: src/main/java/org/folio/dew/repository/BaseFilesStorage.java [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+   * Read all lines from a file
+   *
+   * @param path - the path to the file on S3-compatible storage
+   * @return
+  the lines from the file as a List
+   * @throws IOException - if an I/O error occurs reading from the file
+   */
+  public List<String> readAllLines(String path) throws IOException {
+    try {
+      return lines(path).collect(Collectors.toList());
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dew/repository/BaseFilesStorage.java:390]: Would InputStream be closed properly, that has been created in lines() method?

File: src/test/java/org/folio/dew/BulkEditTest.java [hunk 5]
```diff
@@ -533,33 +533,34 @@ private void verifyFileOutput(JobExecution jobExecution, String output, String e
...
   private JobParameters prepareJobParameters(ExportType exportType, EntityType entityType, IdentifierType identifierType, String path, boolean hasOutcomeFile) {
     Map<String, JobParameter> params = new HashMap<>();
-    if (hasOutcomeFile) {
-      String workDir =
-        System.getProperty("java.io.tmpdir")
-          + File.separator
-          + springApplicationName
-          + File.separator;
+    String jobId = UUID.randomUUID().toString();
+    if (true) {
```
Comment (SerhiiNosko) [inline src/test/java/org/folio/dew/BulkEditTest.java:537]: this if statement can be deleted

File: src/main/java/org/folio/dew/batch/marc/MarcWriter.java [hunk 6]
```diff
@@ -1,40 +1,16 @@
...
 
 @Log4j2
-public class MarcWriter extends FlatFileItemWriter<Record> {
-
-  public MarcWriter(String tempOutputFilePath) {
-    setAppendAllowed(true);
-    setResource(new FileSystemResource(tempOutputFilePath));
+public class MarcWriter extends AbstractStorageStreamWriter<Record, LocalFilesStorage> {
+  public MarcWriter(String tempOutputFilePath, LocalFilesStorage localFilesStorage) {
+    super(tempOutputFilePath, localFilesStorage);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/dew/batch/marc/MarcWriter.java:12]: Basically super invokes parent constructor that does not set any fields. As option we can impement this empty constructor and setResource() call will not be needed

File: src/main/java/org/folio/dew/controller/BulkEditController.java (hunk 1 above)
Comment (khandramai) [reply src/main/java/org/folio/dew/controller/BulkEditController.java:480]: This inputReader is used below in the CSVReader that is in the try-with-resources. So I assumed it should be closed automatically together with CSVReader. But JIC I'll introduce it in the try-with-resources. Thanks.

File: src/test/java/org/folio/dew/BulkEditTest.java (hunk 5 above)
Comment (khandramai) [reply src/test/java/org/folio/dew/BulkEditTest.java:537]: Good catch. Legacy construction.

**Review by SerhiiNosko — APPROVED** [review]

File: src/main/java/org/folio/dew/batch/bursarfeesfines/service/BursarWriterBuilder.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+			}
+		}
+		writer.setLineAggregator(this.lineAggregator);
+		writer.setLineSeparator(this.lineSeparator);
+		writer.setResource(this.resource);
+    writer.setLocalFilesStorage(this.localFilesStorage);
+
+		return writer;
+	}
+}
```
Comment (obozhko-folio) [inline src/main/java/org/folio/dew/batch/bursarfeesfines/service/BursarWriterBuilder.java:182]: Please fix indentation in this file.

**Review by obozhko-folio — APPROVED** [review]

**Review by grabsefx — APPROVED** [review]

**Pull request merged**